### PR TITLE
Add close button to notifications

### PR DESCRIPTION
**Pull request by henriquehbr**
This PR aims to implement a bunch of small additions, most of them are refactorings:

- Add a close button to the notification baloons that can be toggled on/off
> image
- Add a `example` directory in order to test and preview the component
- Convert both `Notifications.svelte` and `store.js` functions to arrow functions
- Bump all dependencies

All the changes were manually tested multiple times in order to assure that they won't break or conflict with existing features

### CODE REVIEW

File: src/Notifications.svelte [hunk 1]
```diff
@@ -168,51 +171,37 @@
...
 			width: '100%'
-		}, ...toasts];
-		count = count + 1
+		}, ...toasts]
+		count += 1
   }
   
   unsubscribe = notification.subscribe(value => {
-    if (!value) { return }
+    if (!value) return
```
Comment (antony) [inline src/Notifications.svelte:199]: I'm not a fan of braceless code, even for single lines, however, I'll add linting to the project to save bike-shedding :)

File: src/Notifications.svelte [hunk 2]
```diff
@@ -168,51 +171,37 @@
...
-			return (y*value)/100;
-		}
-		
-		return {
-			delay,
-			duration,
-			css: t => `opacity: ${(t-.7) * 1}; transform-origin: top right;`
-		}
-	}
+	const animateOut = (node, { delay = 0, duration = 1000 }) => ({
```
Comment (antony) [inline src/Notifications.svelte:180]: weird, I had no idea we didn't use this. nice touch.

File: src/Notifications.svelte [hunk 3]
```diff
@@ -168,51 +171,37 @@
   }
 
   export let timeout = 3000
+  export let closeable = true
```
Comment (antony) [inline src/Notifications.svelte:174]: I wonder if the default should be false...

File: src/notifier.js [hunk 4]
```diff
@@ -1,21 +1,16 @@
 import { notification } from './store.js'
 
-export function send (message, type = 'default', timeout) {
+export const send = (message, type = 'default', timeout) =>
```
Comment (antony) [inline src/notifier.js:3]: We're setting a default and not allowing the possibility of overriding it on a per-notification basis. I think we should add this as a parameter to these.

In fact since we now have multiple options this might be better as:

`msg, options` where options contains `timeout` and `closeable`

**Review by antony — CHANGES_REQUESTED** [review]
This is a great PR! I've suggested some minor changes, let me know what you think.

File: src/Notifications.svelte (hunk 2 above)
Comment (henriquehbr) [reply src/Notifications.svelte:180]: VSCode syntax highlighting really comes in handy these moments 😁

File: src/notifier.js (hunk 4 above)
Comment (henriquehbr) [reply src/notifier.js:3]: That's something that i also felt was a bit off, but wasn't sure whether should be the default behavior or not

Taking into consideration that the options are going to be merged into a single parameter, should we keep all four mode functions (danger, warning, info and success), or may i refactor this file to a single `send` function, in a way the notification type will be passed to the options object:

```js
export const send = (message, options) =>
  notification.set({ message, options })
```
is called as:

```svelte
<button on:click={() => notifier.send('hello', { type: 'danger' })}>Notify</button>
```

or, we can keep all the four object modes, and just adapt their syntax to the `send` parameters changes, like the example below:

```js
export const danger = (msg, options) =>
  send(msg, { type: 'danger', ...options })

export const warning = (msg, options) =>
  send(msg, { type: 'warning', ...options })

export const info = (msg, options) =>
  send(msg, { type: 'info', ...options })

export const success = (msg, options) =>
  send(msg, { type: 'success', ...options })
```

by following this way, we keep the process of calling notifications basically unchanged, but, i personally think that by adhering to this, it will come with a bit of code duplication as a trade-off, but that's just me, choices like these are up to you :)

Comment (antony) [reply src/notifier.js:3]: I would suggest we keep the existing ones, since:

```js
notification.danger('foo')
```

is far more readable and easier to type than:

```js
notifications.send('foo', { type: 'danger' })
```

**Comment by henriquehbr** [comment]
Just pushed the last changes for overriding the close button setting per-notification, and also rewritten some details in the documentation to match the new changes on the notification API, if something seems off or out of place, let me know and i'll take a look at it as soon as possible

Also, one last thing if you don't mind, could you label this PR as `hacktoberfest-accepted` please, It would really help me if this contribution would be counting towards the event goal :)

**Comment by sallaben** [comment]
@antony

**Comment by antony** [comment]
I'll try to review and merge this as soon as possible! Sorry about the delay!

File: README.md [hunk 5]
```diff
@@ -126,13 +126,13 @@ You can set a timeout per message
 import { NotificationDisplay, notifier } from '@beyonk/svelte-notifications'
 
 function someFunction () {
-  notifier.success('Notifications work!', 5000) // built in theme
-  notifier.send('custom-theme', 'Notifications work!', 5000) // custom theme
+  notifier.success('Notifications work!', { duration: 5000 }) // built in theme
```
Comment (antony) [inline README.md:129]: Unfortunately this significantly changes the core API. I like the idea of an options object in principle, but this decision should be made in a completely separate PR which only modifies the API, not muddled in among some refactorings and a new feature.

File: README.md [hunk 6]
```diff
@@ -126,13 +126,13 @@ You can set a timeout per message
...
+  notifier.send('Notifications work!', { type: 'custom-theme', duration: 5000 }) // custom theme
 }
 </script>
 ```
 
 ## Credits
 
-* Original code by [Antony Jones](https://github.com/antony)
-* Animation and performance improvements by jg.
+- Original code by [Antony Jones](https://github.com/antony)
```
Comment (antony) [inline README.md:137]: This just clutters the PR and detracts from the feature you've added.

File: example/App.svelte [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<script>
+  import { NotificationDisplay, notifier } from '../src'
```
Comment (antony) [inline example/App.svelte:2]: I wasn't able to get the demo app to run. Is it worth adding a readme to tell people how to run it?

File: package.json [hunk 8]
```diff
@@ -10,13 +10,13 @@
...
-    "@rollup/plugin-commonjs": "^14.0.0",
-    "@rollup/plugin-node-resolve": "^8.4.0",
-    "rollup": "^1.16.2",
-    "rollup-plugin-postcss": "^2.0.3",
-    "rollup-plugin-serve": "^1.0.1",
-    "rollup-plugin-svelte": "^5.1.0",
-    "svelte": "^3.24.1"
+    "@rollup/plugin-commonjs": "^15.1.0",
+    "@rollup/plugin-node-resolve": "^9.0.0",
+    "rollup": "^2.29.0",
```
Comment (antony) [inline package.json:15]: Upgrading all the libraries as part of a PR for a feature isn't advisable. It's a great idea to do this, but it should be in an isolated PR on its own.

File: rollup.config.js [hunk 9]
```diff
@@ -6,33 +6,33 @@ import commonjs from '@rollup/plugin-commonjs'
...
 }
 
 export default {
-	input: 'src/Notifications.svelte',
-	output: [
-		{ file: pkg.module, 'format': 'es' },
-		{ file: pkg.main, 'format': 'umd', name }
-	],
-	plugins
+  input: 'src/Notifications.svelte',
```
Comment (antony) [inline rollup.config.js:32]: It's best to keep formatting changes out of PRs as it makes it hard to determine what the real changes are.

File: src/Notifications.svelte [hunk 10]
```diff
@@ -3,7 +3,12 @@
 		<li class="toast" style="background: {toast.background};" out:animateOut>
 			<div class="content">
 				{toast.msg}
-			</div>
+      </div>
+      {#if toast.closeable}
```
Comment (antony) [inline src/Notifications.svelte:7]: Looks like a good addition - I think closable toasts should probably be allowed without  a timeout too, so that they **have** to be closed. Not sure how we'd express that in the API.

File: src/Notifications.svelte [hunk 11]
```diff
@@ -168,51 +171,38 @@
...
+  export let closeable = false
 
 	let count = 0
-	let toasts = [ ]
+	let toasts = []
   let unsubscribe
 
-	function animateOut(node, { delay = 0, duration = 1000 }) {
-		function vhTOpx (value) {
-			var w = window,
```
Comment (antony) [inline src/Notifications.svelte:178]: weird, not sure why this dead code was in there!

**Review by antony — CHANGES_REQUESTED** [review]
Apologies it took me nearly a year to review this! I've been a little busy, shall we say.

I think the changes are great but there are 3 or 4 PRs in here rather than one, which makes it hard to review + if I disagree with one part it blocks me merging the whole lot.

I don't know if you're still invested in getting these changes in, but if so, I'd suggest raising the new feature as a PR on its own. We should probably make the timeout optional if you can close notifications.

**Comment by henriquehbr** [comment]
Thanks for your feedback @antony, i'll look towards it maybe later today or tomorrow, seems like a bunch of stuff needs to be improved, anyways, this was made almost an year ago (surely i wasn't the best open-source contributor back then 😆 )

**Comment by antony** [comment]
@henriquehbr no worries. Closing this, and I'll implement some of your ideas separately :)
